src/inabox/inabox-viewport.js

```diff
@@ -0,0 +1,192 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {viewerForDoc} from '../viewer';
+import {Viewport, ViewportBindingDef} from '../service/viewport-impl';
+import {getServiceForDoc} from '../service';
+import {resourcesForDoc} from '../../src/resources';
+import {layoutRectLtwh} from '../layout-rect';
+import {Observable} from '../observable';
+import {MessageType} from '../../src/3p-frame';
+import {IframeMessagingClient} from '../../3p/iframe-messaging-client';
+import {dev} from '../log';
+
+/** @const {string} */
+const TAG = 'inabox-viewport';
+
+/**
+ * Implementation of ViewportBindingDef that works inside an non-scrollable
+ * iframe by listening to host doc for position and resize updates.
+ *
+ * @implements {ViewportBindingDef}
+ */
+export class ViewportBindingInabox {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private @const {!Observable} */
+    this.scrollObservable_ = new Observable();
+
+    /** @private @const {!Observable} */
+    this.resizeObservable_ = new Observable();
+
+    // Before hearing from host doc, we're blind about the viewport position
+    // and iframe position. Positive width/height ensure a render start.
+    /** @private {!../layout-rect.LayoutRectDef} */
+    this.viewportRect_ = layoutRectLtwh(0, 0, 10, 10);
+
+    // To not trigger amp-analytics visibility immediately,
+    // we assume we're right below the fold.
+    /** @private {!../layout-rect.LayoutRectDef} */
+    this.selfRect_ = layoutRectLtwh(0, 10, 10, 10);
+
+    /** @private @const {!IframeMessagingClient} */
+    this.iframeClient_ = new IframeMessagingClient(win);
+    this.iframeClient_.setSentinel(getRandom(win));
+
+    // Bet the top window is the scrollable window and loads host script.
+    // TODOs:
+    // 1) check window ancestor origin, if the top window is in same origin,
+    // don't bother to use post messages.
+    // 2) broadcast the request
+    this.iframeClient_.setHostWindow(win.top);
+
+    dev().fine(TAG, 'initialized inabox viewport');
+  }
+
+  /** @override */
+  connect() {
+    this.iframeClient_.makeRequest(
+        MessageType.SEND_POSITIONS, MessageType.POSITION,
+        data => {
+          dev().fine(TAG, 'Position changed: ', data);
+          const oldViewportRect = this.viewportRect_;
+          const oldSelfRect = this.selfRect_;
+          this.viewportRect_ = data.viewport;
+          this.selfRect_ = data.target;
+          if (isMoved(this.selfRect_, oldSelfRect)) {
+            // Remeasure all AMP elements once iframe position is changed.
+            // Because all layout boxes are calculated relatively to the
+            // iframe position.
+            this.remeasureAllElements_();
```

Comment:
I am not sure about this. When an amp element is inabox. its `layoutBox` value should be relative to the box. Changing the position of the box shouldn't affect their cached layoutBox value?